sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java

```diff
@@ -140,23 +143,44 @@ public VectorizedColumnReader(
     this.writerVersion = writerVersion;
   }
 
-  private boolean isLazyDecodingSupported(PrimitiveType.PrimitiveTypeName typeName) {
+  private boolean isLazyDecodingSupported(PrimitiveType.PrimitiveTypeName typeName,
+                                          DataType sparkType) {
+    // Don't use lazy dictionary decoding if the column needs extra processing: upcasting or date /
+    // decimal scale rebasing.
     return switch (typeName) {
-      case INT32 ->
-        !(logicalTypeAnnotation instanceof DateLogicalTypeAnnotation) ||
-          "CORRECTED".equals(datetimeRebaseMode);
+      case INT32 -> {
+        boolean needsUpcast = sparkType == LongType || sparkType == TimestampNTZType ||
+                !DecimalType.is32BitDecimalType(sparkType);
+        boolean needsRebase = logicalTypeAnnotation instanceof DateLogicalTypeAnnotation &&
+          !"CORRECTED".equals(datetimeRebaseMode);
+        yield !needsUpcast && !needsRebase && !needsDecimalScaleRebase(sparkType);
+      }
       case INT64 -> {
-        if (updaterFactory.isTimestampTypeMatched(TimeUnit.MICROS)) {
-          yield "CORRECTED".equals(datetimeRebaseMode);
-        } else {
-          yield !updaterFactory.isTimestampTypeMatched(TimeUnit.MILLIS);
-        }
+        boolean needsUpcast = !DecimalType.is64BitDecimalType(sparkType) ||
+                updaterFactory.isTimestampTypeMatched(TimeUnit.MILLIS);
+        boolean needsRebase = updaterFactory.isTimestampTypeMatched(TimeUnit.MICROS) &&
+          !"CORRECTED".equals(datetimeRebaseMode);
+        yield !needsUpcast && !needsRebase && !needsDecimalScaleRebase(sparkType);
       }
-      case FLOAT, DOUBLE, BINARY -> true;
+      case FLOAT -> sparkType == FloatType;
+      case DOUBLE, BINARY -> !needsDecimalScaleRebase(sparkType);
       default -> false;
     };
   }
 
+  /**
+   * Returns whether the Parquet type of this column and the given spark type are two decimal types
+   * with different scales.
```

Comment:
I may missed some details. Where is the code to handle decimal scale difference during parquet scan?

Reply:
That shouldn't be part of this change I removed it. I started adding support for reading decimals using a larger scale and precision but it started growing too large so I cut it out, I plan on following with a separate PR for it.